### DISCUSSION TITLE
Packaging werf for openSUSE: build fails with "pkg/container_backend/instruction/run.go:39:22: undefined: instructions.GetSecurity"

### DISCUSSION THREAD

**Discussion by johanneskastl** [werf/werf#5875] [General]
Hi,

I am trying to package werf for openSUSE before playing around with it. But the build fails with the following error:

```
[    9s] + go build -mod=vendor -buildmode=pie '-ldflags=-X github.com/werf/werf/pkg/werf.Version=1.2.270' -o bin/werf ./cmd/werf
[   16s] # github.com/werf/werf/pkg/container_backend/instruction
[   16s] pkg/container_backend/instruction/run.go:39:22: undefined: instructions.GetSecurity
[   20s] error: Bad exit status from /var/tmp/rpm-tmp.PZ3Ws1 (%build)
```

I could workaround the missing dependencies for compilation, but I am not familiar enough with golang to get this solved myself. Looks to me like I have a dependency that is too old, but which could this be?

Any hints are highly appreciated.

**Comment by distorhead** [comment]
@johanneskastl Hello! Thanks for your interest in werf.

This error occurs because we need to pass some special go-build-tags to enable some library functions used by the werf (dfrunsecurity build tag in this concrete case). But there are more build tags and all of them are specified in the project taskfile (https://taskfile.dev/, sort of makefile): https://github.com/werf/werf/blob/main/Taskfile.dist.yaml#L22.

Important thing to know is that build for linux should enable CGO and list of required go-build-tags, which is:

```
  cgoTags: "dfrunsecurity dfrunnetwork dfrunmount dfssh containers_image_openpgp osusergo exclude_graphdriver_devicemapper netgo no_devmapper static_build"
  cgoDevLDFlags: "-linkmode external -extldflags=-static"
  cgoReleaseLDFlags: "-linkmode external -extldflags=-static -s -w -X github.com/werf/werf/pkg/werf.Version={{.version}}"
```

The build command will be:

```
      - go build -o "{{.outputDir | default (printf "./dist/%s/%s-%s/bin" .version .targetOS .targetArch)}}/werf{{if (eq .targetOS "windows")}}.exe{{end}}"
        -compiler gc -ldflags="{{.cgoReleaseLDFlags}}" -tags="{{.cgoTags}}" {{.CLI_ARGS}} {{.pkg | default .package}}
```

The most preferred way to build werf as we build it is to use task `build:dist:all` or subordinary tasks like `build:dist:linux:amd64` and `build:dist:linux:arm64`: https://github.com/werf/werf/blob/main/Taskfile.dist.yaml#L90.
But you can call go build by yourself passing all needed flags.

**Reply by johanneskastl** [reply]
Thanks, that solved it. I missed the tags....

Now it is building and I can start exploring...
